File: apps/api/src/competehub_api/schemas/calendar.py
from __future__ import annotations

from datetime import date

from marshmallow import Schema, ValidationError, fields, validate, validates_schema

from competehub_api.schemas.common import UtcDateTime

CALENDAR_DATE_MIN = date(1, 1, 2)
CALENDAR_DATE_MAX = date(9999, 12, 30)


class CalendarQuerySchema(Schema):
    date_from = fields.Date(
        required=True,
        data_key="from",
        validate=validate.Range(min=CALENDAR_DATE_MIN, max=CALENDAR_DATE_MAX),
    )
    date_to = fields.Date(
        required=True,
        data_key="to",
        validate=validate.Range(min=CALENDAR_DATE_MIN, max=CALENDAR_DATE_MAX),
    )
    view = fields.String(
        required=True,
        validate=validate.OneOf(["month", "week", "list"]),
    )

    @validates_schema
    def validate_range(self, data, **kwargs):
        if data.get("date_from") and data.get("date_to"):
            if data["date_from"] > data["date_to"]:
                raise ValidationError(
                    "Calendar end must not be before calendar start.",
                    field_name="to",
                )


class CalendarRangeSchema(Schema):
    date_from = fields.Date(attribute="from", data_key="from")
    date_to = fields.Date(attribute="to", data_key="to")
    view = fields.String()
    time_zone = fields.String()


class CalendarItemSchema(Schema):
    competition_id = fields.Integer()
    competition_title = fields.String()
    detail_url = fields.String(allow_none=True)
    lifecycle_status = fields.String()
    target_available = fields.Boolean()
    stage_id = fields.Integer(allow_none=True)
    stage_label = fields.String(allow_none=True)
    stage_order = fields.Integer(allow_none=True)
    stage_type = fields.String(allow_none=True)
    is_current_stage = fields.Boolean()
    node_snapshot_id = fields.Integer()
    logical_node_key = fields.String(allow_none=True)
    node_revision = fields.Integer()
    node_type = fields.String()
    description = fields.String(allow_none=True)
    occurs_at = UtcDateTime()
    prominence = fields.String()
    pair_kind = fields.String(allow_none=True)
    pair_role = fields.String(allow_none=True)


class CalendarPayloadSchema(Schema):
    range = fields.Nested(CalendarRangeSchema())
    items = fields.List(fields.Nested(CalendarItemSchema()))


calendar_query_schema = CalendarQuerySchema()
calendar_payload_schema = CalendarPayloadSchema()

File: apps/web/src/types/calendar.ts
export type CalendarView = 'month' | 'week' | 'list'

export interface CalendarRange {
  from: string
  to: string
  time_zone: 'Asia/Shanghai'
  view: CalendarView
}

export interface CalendarItem {
  competition_id: number
  competition_title: string
  detail_url: string | null
  lifecycle_status: string
  target_available: boolean
  stage_id: number | null
  stage_label: string | null
  stage_order: number | null
  stage_type: string | null
  is_current_stage: boolean
  node_snapshot_id: number
  logical_node_key: string | null
  node_revision: number
  node_type: string
  description: string | null
  occurs_at: string
  prominence: 'primary' | 'secondary'
  pair_kind: 'registration' | 'competition' | null
  pair_role: 'start' | 'deadline' | 'end' | null
}

export interface CalendarPayload {
  range: CalendarRange
  items: CalendarItem[]
}

export interface CalendarQuery {
  from: string
  to: string
  view: CalendarView
}
